File: apps/api_py/src/linkbook/actions/queue.py
"""§2.5 — 30s soft-undo for email/Slack actions.

In-process v1: an asyncio task per row, persisted to actions.queued_until
so it survives restart. Race-free CAS on actions.status.
"""

from __future__ import annotations

import asyncio
import os
import random
from datetime import datetime, timedelta

from sqlalchemy import and_, select, update
from sqlalchemy.orm import Session

from ..config import AppConfig
from ..db.models import Action, AuditEvent
from ..lib.log import AppLogger

QUEUEABLE = {"invoice.remind", "email.send_draft", "time.self_nudge"}

_TIMERS: dict[str, asyncio.Task[None]] = {}


def is_queueable(type_: str) -> bool:
    return type_ in QUEUEABLE


def get_send_delay_ms() -> int:
    raw = os.environ.get("SEND_DELAY_MS")
    if raw:
        try:
            n = int(raw)
            if n > 0:
                return n
        except ValueError:
            pass
    return 30_000


def cas_status(
    db: Session, action_id: str, *, frm: str, to: str, **fields: object
) -> int:
    """Compare-and-swap status. Returns rowcount (0 or 1)."""
    stmt = (
        update(Action)
        .where(and_(Action.id == action_id, Action.status == frm))
        .values(status=to, updated_at=datetime.utcnow(), **fields)
    )
    res = db.execute(stmt)
    db.commit()
    return res.rowcount or 0


async def enqueue_send_delay(db: Session, action_id: str) -> None:
    delay_ms = get_send_delay_ms()
    queued_until = datetime.utcnow() + timedelta(milliseconds=delay_ms)

    changed = cas_status(db, action_id, frm="drafted", to="queued_30s", queued_until=queued_until)
    if changed == 0:
        cur = db.get(Action, action_id)
        if cur is None:
            raise RuntimeError("action not found")
        if cur.status != "queued_30s":
            raise RuntimeError("action not in drafted state")
    a = db.get(Action, action_id)
    if a is None:
        return
    db.add(
        AuditEvent(
            actor="system",
            action_id=a.id,
            originating_event_id=a.originating_event_id,
            kind="action.queued",
            idempotency_key=a.idempotency_key,
            subject_ref=a.subject_ref,
        )
    )
    db.commit()
    _arm_timer(action_id, delay_ms)


def _arm_timer(action_id: str, ms: int) -> None:
    """Schedule an async task to fire when the timer elapses."""
    loop = asyncio.get_event_loop()

    async def fire() -> None:
        await asyncio.sleep(max(0, ms / 1000))
        _TIMERS.pop(action_id, None)
        # Open a fresh session — the original db session may be closed.
        from ..db import open_session
        from ..config import load_config
        from ..lib.log import app_logger

        cfg = load_config()
        log = app_logger()
        db = open_session()
        try:
            won = cas_status(
                db,
                action_id,
                frm="queued_30s",
                to="executing",
                executed_at=datetime.utcnow(),
            )
            if won == 0:
                log.info({"action_id": action_id}, "queue: timer fired but action already cancelled")
                return
            a = db.get(Action, action_id)
            if a is None:
                return
            db.add(
                AuditEvent(
                    actor="system",
                    action_id=a.id,
                    originating_event_id=a.originating_event_id,
                    kind="action.executing",
                    idempotency_key=a.idempotency_key,
                    subject_ref=a.subject_ref,
                    note="queue timer elapsed",
                )
            )
            db.commit()
            from .execute import ExecuteOptions, execute_action

            await execute_action(cfg, db, log, a, ExecuteOptions(already_executing=True))
        except Exception as e:  # noqa: BLE001
            log.error({"err": str(e), "action_id": action_id}, "queue dispatch failed")
        finally:
            db.close()

    task = loop.create_task(fire())
    _TIMERS[action_id] = task


async def cancel_send_delay(db: Session, action_id: str) -> bool:
    """§2.5 soft-undo: clicked Undo within the 30s window.

    Returns the action to `drafted` so the user can approve again or reject.
    The integration call never fired, so there's nothing to compensate.

    State transition: queued_30s → drafted (with queued_until cleared).
    """
    t = _TIMERS.pop(action_id, None)
    if t is not None:
        t.cancel()
    changed = cas_status(
        db, action_id, frm="queued_30s", to="drafted", queued_until=None
    )
    if changed == 0:
        return False
    a = db.get(Action, action_id)
    if a is None:
        return False
    db.add(
        AuditEvent(
            actor="user",
            action_id=a.id,
            originating_event_id=a.originating_event_id,
            kind="action.unqueued",
            idempotency_key=a.idempotency_key,
            subject_ref=a.subject_ref,
            note="soft-undo (30s window) — returned to drafted",
        )
    )
    db.commit()
    return True


async def recover_queue_on_boot(db: Session, log: AppLogger) -> None:
    """Boot recovery: fire past-deadline rows with random jitter; re-arm
    timers for the rest. Persistent state across restarts (§2.5)."""
    rows = db.execute(select(Action).where(Action.status == "queued_30s")).scalars().all()
    for a in rows:
        remaining_ms = (
            int((a.queued_until - datetime.utcnow()).total_seconds() * 1000)
            if a.queued_until is not None
            else 0
        )
        arm_after = remaining_ms if remaining_ms > 0 else random.randint(0, 500)
        _arm_timer(a.id, arm_after)
    if rows:
        log.info({"count": len(rows)}, "queue: recovered queued_30s actions on boot")


def clear_all_timers() -> None:
    for t in list(_TIMERS.values()):
        t.cancel()
    _TIMERS.clear()

File: apps/api/src/actions/queue.ts
// §2.5 — 30s soft-undo for email/Slack actions.
// In-process v1: a setTimeout per queued action.
// Restart recovery: on boot, drain queued_30s rows whose timer has already
// elapsed and re-arm setTimeout for the rest. (Bridges the gap until the
// Conductor workflow lands.)
//
// Race-free: timer fire and undo both use a CAS update on the status
// column so only one can win.

import type { ActionType } from '@linkbook/types';
import type { AppCtx } from '../lib/principal-context.js';
import { findActionById, writeAudit } from '../db/actions-repo.js';
import { executeAction } from './execute.js';
import { schema } from '@linkbook/db';
import { and, eq, sql } from 'drizzle-orm';

const TIMERS = new Map<string, NodeJS.Timeout>();

const QUEUEABLE: readonly ActionType[] = ['invoice.remind', 'email.send_draft', 'time.self_nudge'];

export function isQueueable(type: ActionType): boolean {
  return QUEUEABLE.includes(type);
}

// Configurable via env so tests can run against a fast queue.
export function getSendDelayMs(env: NodeJS.ProcessEnv = process.env): number {
  const raw = env['SEND_DELAY_MS'];
  if (!raw) return 30_000;
  const n = Number(raw);
  return Number.isFinite(n) && n > 0 ? n : 30_000;
}

// Race-free CAS: returns the number of rows affected (0 or 1). better-sqlite3
// reliably returns { changes }, but we double-check with a post-update read
// in case a future driver swap changes that contract.
function casStatus(
  ctx: AppCtx,
  action_id: string,
  from: string,
  to: string,
  patch?: Partial<typeof schema.actions.$inferInsert>,
): number {
  const result = ctx.db
    .update(schema.actions)
    .set({ ...patch, status: to as never, updated_at: new Date() })
    .where(and(eq(schema.actions.id, action_id), eq(schema.actions.status, from as never)))
    .run();
  if (typeof result.changes === 'number') return result.changes;
  // fallback: confirm by reading
  const row = ctx.db.select().from(schema.actions).where(eq(schema.actions.id, action_id)).all()[0];
  return row && row.status === to ? 1 : 0;
}

export async function enqueueSendDelay(ctx: AppCtx, action_id: string): Promise<void> {
  const delay = getSendDelayMs();
  const queued_until = new Date(Date.now() + delay);
  // CAS from drafted → queued_30s; protects against double-approval.
  const changed = casStatus(ctx, action_id, 'drafted', 'queued_30s', { queued_until });
  if (changed === 0) {
    // already moved on
    const cur = await findActionById(ctx, action_id);
    if (cur && cur.status === 'queued_30s') {
      // already queued; just (re-)arm the timer
      armTimer(ctx, action_id, queued_until.getTime() - Date.now());
      return;
    }
    throw new Error('action not in drafted state');
  }
  const a = await findActionById(ctx, action_id);
  await writeAudit(ctx, {
    actor: 'system',
    action_id,
    kind: 'action.queued',
    idempotency_key: a?.idempotency_key ?? null,
    subject_ref: a?.subject_ref ?? '',
  });
  armTimer(ctx, action_id, delay);
}

function armTimer(ctx: AppCtx, action_id: string, ms: number): void {
  const t = setTimeout(() => {
    TIMERS.delete(action_id);
    // CAS from queued_30s → executing. If user already cancelled, we lose.
    const won = casStatus(ctx, action_id, 'queued_30s', 'executing', { executed_at: new Date() });
    if (won === 0) {
      ctx.log.info({ action_id }, 'queue: timer fired but action already cancelled');
      return;
    }
    findActionById(ctx, action_id)
      .then(async (a) => {
        if (!a) return;
        // §2.6 audit trail for the queued_30s → executing transition.
        await writeAudit(ctx, {
          actor: 'system',
          action_id,
          originating_event_id: a.originating_event_id,
          kind: 'action.executing',
          idempotency_key: a.idempotency_key,
          subject_ref: a.subject_ref,
          note: 'queue timer elapsed',
        });
        // alreadyExecuting=true so executeAction doesn't redundantly flip
        // status (which would clobber any concurrent terminal state).
        await executeAction(ctx, a, { alreadyExecuting: true });
      })
      .catch((e) => ctx.log.error({ err: e, action_id }, 'queue dispatch failed'));
  }, Math.max(0, ms));
  t.unref?.();
  TIMERS.set(action_id, t);
}

export async function cancelSendDelay(ctx: AppCtx, action_id: string): Promise<boolean> {
  // Cancel the in-process timer first (best-effort) then CAS the row.
  const t = TIMERS.get(action_id);
  if (t) {
    clearTimeout(t);
    TIMERS.delete(action_id);
  }
  const changed = casStatus(ctx, action_id, 'queued_30s', 'cancelled');
  if (changed === 0) return false;
  const a = await findActionById(ctx, action_id);
  await writeAudit(ctx, {
    actor: 'user',
    action_id,
    kind: 'action.cancelled',
    idempotency_key: a?.idempotency_key ?? null,
    subject_ref: a?.subject_ref ?? '',
    note: 'soft-undo (30s window)',
  });
  return true;
}

// Boot-time recovery — call from server.ts after migrations apply.
// Past-deadline rows fire with random jitter (0–500ms) to avoid a
// thundering herd against integrations after a long downtime; future
// rows arm at their remaining time.
export async function recoverQueueOnBoot(ctx: AppCtx): Promise<void> {
  const queued = ctx.db
    .select()
    .from(schema.actions)
    .where(eq(schema.actions.status, 'queued_30s'))
    .all();
  for (const a of queued) {
    const remaining = a.queued_until ? a.queued_until.getTime() - Date.now() : 0;
    const armAfter = remaining > 0 ? remaining : Math.floor(Math.random() * 500);
    armTimer(ctx, a.id, armAfter);
  }
  if (queued.length > 0) {
    ctx.log.info({ count: queued.length }, 'queue: recovered queued_30s actions on boot');
  }
}

export function _clearAllTimers(): void {
  for (const t of TIMERS.values()) clearTimeout(t);
  TIMERS.clear();
}

// kept for compatibility with old callers; readers use getSendDelayMs() now
export const SEND_DELAY_MS = 30_000;
